crowbar_framework/app/models/node_object.rb

```diff
@@ -706,58 +706,40 @@ def crowbar_run_list(*args)
   end
 
   def add_to_run_list(rolename, priority, states = nil)
-    states = ["all"] unless states
-    crowbar["run_list_map"] = {} if crowbar["run_list_map"].nil?
-    val = { "states" => states, "priority" => priority }
-    crowbar["run_list_map"][rolename] = val
-    Rails.logger.debug("crowbar[run_list_map][#{rolename}] = #{val.inspect}")
-    Rails.logger.debug("current state is #{self.crowbar['state']}")
-
-    # only rebuild the run_list if it effects the current state.
-    self.rebuild_run_list if states.include?("all") or states.include?(self.crowbar["state"])
+    # FIXME: Crowbar 4.0: we keep states parameter for compatibility reason; it
+    # should be removed in Crowbar 5.0
+    crowbar["run_list_map"] ||= {}
+    crowbar["run_list_map"][rolename] = { "priority" => priority }
+    rebuild_run_list
   end
 
   def delete_from_run_list(rolename)
-    crowbar["run_list_map"] = {} if crowbar["run_list_map"].nil?
-    crowbar["run_list_map"][rolename] = { "states" => ["all"], "priority" => -1001 } unless crowbar["run_list_map"].nil?
+    crowbar["run_list_map"] ||= {}
+    crowbar["run_list_map"].delete(rolename)
     crowbar_run_list.run_list_items.delete "role[#{rolename}]"
   end
 
   def rebuild_run_list
-    crowbar["run_list_map"] = {} if crowbar["run_list_map"].nil?
+    crowbar["run_list_map"] ||= {}
 
-    # Cull by state
-    map = crowbar["run_list_map"].select do |k, v|
-      v["states"].include?("all") || v["states"].include?(crowbar["state"])
-    end
+    # FIXME: Crowbar 4.0: remove this hack to drop old items, which exists only
+    # to allow upgrade from 3.0 to work; should be removed in Crowbar 5.0
+    crowbar["run_list_map"].delete_if { |k, v| v["priority"] == -1001 }
```

Comment:
MINOR^2: could use `_k` for the unused block variable. normally hound used to complain about that...